File: 016.3SumClosest/sum.cc
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        int ans = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < int(nums.size()) - 2; i++) {
            if (i != 0 && nums[i] == nums[i-1]) {
                continue;
            }

            int l = i + 1, r = int(nums.size()) - 1;
            while (l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                if (sum == target) {
                    return sum;
                }

                if (abs(sum - target) < abs(ans - target)) {
                    ans = sum;
                    while (l < r && nums[l] == nums[l+1]) l++;
                    while (l < r && nums[r] == nums[r-1]) r--;
                }

                if (sum < target) l++;
                else r--;
            }
        }

        return ans;
    }
};

File: 016.3SumClosest/sum.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()

        ans = nums[0] + nums[1] + nums[2]

        for i in range(len(nums) - 2):
            if i != 0 and nums[i] == nums[i-1]:
                continue

            l, r = i+1, len(nums)-1
            while l < r:
                s = nums[l] + nums[r] + nums[i]
                if s == target:
                    return s

                if abs(s - target) <  abs(ans - target):
                    ans = s
                    while l < r and nums[l] == nums[l+1]: l += 1
                    while l < r and nums[r] == nums[r-1]: r -= 1

                if s < target: l += 1
                else: r -= 1

        return ans

if __name__ == '__main__':
    nums = [-1,0,1,1,55]
    target = 3
    print Solution().threeSumClosest(nums, target)
